helix-core/src/main/java/org/apache/helix/model/InstanceConfig.java

```diff
@@ -637,4 +638,19 @@ public static InstanceConfig toInstanceConfig(String instanceId) {
     }
     return config;
   }
+
+  /**
+   * Validate if the topology related settings (Domain or ZoneId) in the given instanceConfig
+   * are valid and align with current clusterConfig.
+   * This function should be called when instance added to cluster or caller updates instanceConfig.
+   *
+   * @throws IllegalArgumentException
+   */
+  public boolean validateTopologySettingInInstanceConfig(ClusterConfig clusterConfig,
+      String instanceName) throws IllegalArgumentException{
```

Comment:
Usually runtime exception is not recommended to put in method signature. Having it in java doc is good enough.